File: dotnet-desktop-guide/net/winforms/controls-design/snippets/usercontrol-guidelines/csharp/UserControl1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace WinFormsProject1
{
    public partial class UserControl1 : UserControl
    {
        //<property_changed>
        // The event
        public event EventHandler AllowInteractionChanged;

        // The backing field for the property
        private bool _allowInteraction;

        // The property
        public bool AllowInteraction
        {
            get => _allowInteraction;
            set
            {
                // When the value has changed, call the method to raise the event
                if (_allowInteraction != value)
                {
                    _allowInteraction = value;
                    OnAllowInteractionChanged();
                }
            }
        }

        // Raises the event
        public virtual void OnAllowInteractionChanged() =>
            AllowInteractionChanged?.Invoke(this, EventArgs.Empty);
        //</property_changed>


        //<eventargs>
        public class ProgressReportEventArgs : EventArgs
        {
            public readonly int Value;
            public readonly int Maximum;

            public ProgressReportEventArgs(int value, int maximum) =>
                (Value, Maximum) = (value, maximum);
        }

        public event EventHandler<ProgressReportEventArgs> ProgressChanged;

        public virtual void OnProgressChanged(int value, int maximum) =>
            ProgressChanged?.Invoke(this, new ProgressReportEventArgs(value, maximum));
        //</eventargs>





        public UserControl1()
        {
            InitializeComponent();
        }


    }
}

File: dotnet-desktop-guide/net/winforms/controls-design/snippets/usercontrol-guidelines/vb/UserControl1.vb
Public Class UserControl1

    '<property_changed>
    'The event
    Public Event AllowInteractionChanged As EventHandler

    'The backing field for the property
    Private _allowInteraction As Boolean

    'The property
    Public Property AllowInteraction() As Boolean
        Get
            Return _allowInteraction
        End Get
        Set(value As Boolean)
            _allowInteraction = value
            OnAllowInteractionChanged()
        End Set
    End Property

    'Raises the event
    Public Overridable Sub OnAllowInteractionChanged()
        RaiseEvent AllowInteractionChanged(Me, EventArgs.Empty)
    End Sub
    '</property_changed>

    '<eventargs>
    Public Class ProgressReportEventArgs
        Inherits EventArgs

        Public ReadOnly Value As Integer
        Public ReadOnly Maximum As Integer

        Public Sub New(value As Integer, maximum As Integer)
            Me.Value = value
            Me.Maximum = maximum
        End Sub
    End Class

    Public Event ProgressChanged As EventHandler(Of ProgressReportEventArgs)

    Public Overridable Sub OnProgressChanged(value As Integer, maximum As Integer)
        RaiseEvent ProgressChanged(Me, New ProgressReportEventArgs(value, maximum))
    End Sub
    '</eventargs>
End Class
